### DISCUSSION TITLE
Built-In/Global Variables

### DISCUSSION THREAD

**Discussion by LukeTillman** [perses/perses#226] [Design Discussions]
One thing that's currently preventing the Node Exporter dashboard from rendering completely is missing variables, in particular, missing "built-in" or "global" variable values. Grafana supports two different sets of these:

1. Truly [global variables](https://grafana.com/docs/grafana/v7.5/variables/variable-types/global-variables/) that are available for all data sources.
2. Variables that are built-in to specific data sources, for example, the [Prometheus Data Source](https://grafana.com/docs/grafana/v7.5/datasources/prometheus/#templating).

All of these variables follow the `$__name` naming convention and provide access to different pieces of the Grafana state inside of templated expressions (typically query expressions, but templating is also sometimes available in chart formatting/options). Here's a quick summary of all the supported variables in those docs links above:

### Global
1. `$__dashboard`: the name of the current dashboard
1. `$__from` and `$__to`: current time range start and end as epoch milliseconds
2. `$__name`: series name or alias (only available in SingleStat panel field options)
3. `$__org`: current org's ID
4. `${__user.[id|login|email]}`: information about the currently logged in user
5. `$__timeFilter`: currently selected time range as an "expression" (e.g. "Last 7 days" becomes `time > now() - 7d`).

### Prometheus
1. `$__interval` and `$__interval_ms`: the step for a Prom range query as both a duration string and in milliseconds (e.g. `60s` and `60000`).
2. `$__range`, `$__range_s`, and `$__range_ms`: the duration between the from/to dates used for a Prom range query as a duration string and in seconds/milliseconds (e.g. `60s`, `60`, `60000`).
3. `$__rate_interval`: interval in seconds for use in `rate()` functions that implements the Prom best practice of "the range in a rate query should be at least four times the scrape interval" (see the linked docs above for calculation details)

Here are some of the questions I have for discussion:

1. What problem are these trying to solve and do we need to solve the same problem in Perses?
2. Is this (built-in/global variables) the best or only way to solve that problem?
3. What do we do with these when converting from Grafana JSON? Are they converted to something else? Do we just not support them? etc.

**Comment by AntoineThebaud** [comment]
I can speak for my case here.. I've gotten used to use `$__interval` in all the rate() calls I'm doing, instead of putting plain duration value, because doing so can lead to "miss" data when visualizing large periods : doing rate[1m] when visualizing 1 month of data leads to ignoring part of the data as there's not enough pixels available to render 1 point every minute.. as the prometheus step parameter is automatically adjusting to this, rate[$__interval] will be like rate[30m] so you won't ignore data. From the [documentation](https://grafana.com/docs/grafana/v7.5/variables/variable-types/global-variables/#__interval) it seems to be the main purpose of this variable. Actually `$__rate_interval` was added since and it's even better because it tackles this problem + the recommendation to have at least 4 points to rate on.

Regarding `$__range` I have personally 2 main usages of it : 
- with pie charts, to render the `<aggregation>` of data over the current range
- within grafana variables, for the case of volatile labels : when displaying 3 hours of data, I want my dropdown to show the label values that existed at some point during this range, not only the instant/current ones.

I really think those built-in Prometheus variables will be must-have in Perses too. I cannot tell for the Global ones as I didn't use them much.

**Comment by Gladorme** [comment]
Hello, I took some time to look at our usage of global variables in Amadeus dashboards.
We are heavily using the following variables: `$__rate_interval`, `$__from` and `$__to`.


## Usage

- `$__rate_interval` is mostly used as Antoine told in his [previous post](https://github.com/perses/perses/discussions/226#discussion-3730199) => Apply best pratices for rate() calls.
- `$__from` and `$__to` are mainly used to build link urls to go to other dashboards with the same time range as the dashboard currently viewed. Example: 
``` json 
"links": [
    {
        "targetBlank": true,
        "title": "[Explore] By namespace",
        "url": "/explore?left={\"datasource\":\"$datasource\",\"queries\":[{\"expr\":\"vector(1) == 1\"}],\"range\":{\"from\":\"$__from\",\"to\":\"$__to\"}}"
    },
    ...
]
```

By curiosity, I quickly checked Grafana most downloaded dashboards. Few dashboards are using global variables:
- `$__dashboard`: **NONE**
- `$__from` and `$__to`: **NONE**
- `$__name`: **NONE**
- `$__org`: **NONE**
- `${__user.[id|login|email]}`: **NONE**
- `$__timeFilter`: **NONE**

- `$__interval` and `$__interval_ms`: alertmanager + redis dashboard
- `$__interval_ms`, `$__range_s`, and `$__range_ms`: **NONE**
- `$__rate_interval`: node-exporter-full

- `$__auto_interval_interval`: argocd + postgresql + elasticsearch (Found this variables in some dashboards, but I didn't find any mention in Grafana doc) 

- No global variables: istio + kube-state-metrics + jvm-micrometer + loki-stack + kubernetes-apiserver + rabbitmq + nginx-ingress-controller + go_processes + coredns + fluentd + prometheus-2.0 + aws-rds + traefik


## User defined variables

It would be nice to allow users to define global variables at global level or project level, like Datadog is proposing with [Template variable](https://docs.datadoghq.com/dashboards/template_variables/).

For example, as of today in our company: anyone can deploy a new k8s cluster with monitoring enabled. In our Grafana dashboards, there is a "cluster" variable computed each time you use them, in order to filter on any k8s cluster where there is the monitoring enabled (prometheus / thanos / ...). For filling this variable, we are using the discovery query: `label_values(prometheus_build_info,cluster)`. It's a query computed thousands of time by day, on hundred of clusters and could be ease with a "static" list. Where this list could be easily updated with a http call  during the process to add new cluster. No need to wait few of seconds for the query to be computed and lessen the computation/bandwidth/... usage.


## Multiple types of variables

In the end, I see three types of "global" variables:
1. Dynamic global variables: the variable value is depending on something or is processing some logic (current timerange, current dashboard, ...)
2. Datasource global variables: a dynamic global variable restricted to a datasource (provided with the datasource plugin?)
3. Static global variables: the variable value can be defined by an user

I/my team think we should implement global variables, they are really useful to be build better dashboards. Next, is to see how they can be implemented (using plugins, ...) :smile:

**Reply by kmalhi** [reply]
One thing I want to discuss is the ability to have truly global user defined variables (static or dynamic). In k8s, there are some standard constructs like cluster, namespace, pod, node etc. These variables are used across a lot of dashboards. Another example would be `interval`.  

AFAIK, datadog and grafana allow dashboard-local user-defined template variables. 

Users have to replicate these user-defined variable definitions when they create a new dashboard. Often they will also name the variable slightly differently i.e. either use upper/lower case for the name or abbreviate it. For example, I have seen the `environment` variable being named as env, Env, Environment, environment etc -- and they all generally mean the same thing i.e a list of environments. 

Secondly, users often get confused when defining template variables. It adds friction when defining queries for a dashboard. As a result, users end up copying an existing dashboard as a starting point for their new dashboard and then start modifying it. This leads to dashboard sprawl as well and finding the single source of truth becomes harder. 

Giving the ability to an admin or super user to define these variables globally might go a long way in helping users with template variables . A user could simply re-use a global user-defined template variable in their dashboard or override it if they want to customize it .

**Reply by Nexucis** [reply]
> One thing I want to discuss is the ability to have truly global user defined variables (static or dynamic).

I guess that's what we are calling now `GlobalVariable` and `ProjectVariable`. That's the same thing than the datasource. You can define them globally or at project level with an overridden logic.

These variables are persisted in the database and loaded at the same time than the dashboard. 

Here an example on what would be the definition of the variables:

```json
{
  "kind": "GlobalVariable"
  "metadata": {
    "name": "namespace"
  },
  "spec": {
    "kind": "ListVariable",
    "spec": {
      "allow_multiple": false,
      "allow_all_value": false,
      "default_value": "demo.do.prometheus.io:9100",
      "plugin": {
         "kind": "PrometheusLabelValuesVariable",
         "spec": {
             "label_name": "instance",
             "matchers": ["up{job=~\"$job\"}"]
         }
      }
    }
  }
}

**Comment by sjcobb** [comment]
I really like the ideas around user defined global-level and project-level variables! Those could almost be their own discussion entirely since the scope is so large. 

For the built-in global variables, agreed we definitely use them a ton at our company too (our PM is compiling some data about exactly which ones are used and how much). Which specific Grafana global variables are you looking into implementing? Also what are the options for syntax, are you thinking sticking with the `$__` prefix?

**Reply by Gladorme** [reply]
I think we would like to implement `$__rate_interval`, `$__from` and `$__to` in priority as we are using them, wdyt @Nexucis?
I think we can keep the `$__` prefix. It's already used today in Perses with the `$__all` variable. Except if there is a reason to change?

**Reply by Nexucis** [reply]
yeah that would be the priority on our side. But it's based on our own analyse on the variable usage.
I guess it won't hurt if we are using the same syntax.

**Reply by Nexucis** [reply]
but this syntax is based on the fact that the variable syntax is the same for everything i.e. `$<variable_name>`. Which is not the case for the moment on our side.

And I really think we shouldn't have a variable syntax that can change depending of the plugin. It's a nice idea, but it's a complex one and from my perspective it's not necessary.
For once I think we can leverage on the Grafana experience. They are managing a lot of different datasource with different query syntax and the variable syntax is the same across every query language. So I think we can access the same thing. It will simplify many things and it won't hurt Perses.

**Reply by sjcobb** [reply]
Wouldn’t it save us work long term though? 

If we let datasources be opinionated about this, we can use more out-of-the-box open source solutions instead of building it all ourselves.

I found the idea that for example, a MySQL datasource could use the `@var_name` syntax really appealing. I’m not sure this is the place to copy Grafana 😬

**Reply by AntoineThebaud** [reply]
> If we let datasources be opinionated about this, we can use more out-of-the-box open source solutions instead of building it all ourselves.

Could you provide more details here?

Plugin-opinionated syntax means bigger learning curves for end users manipulating multiple types of datasources. Maybe that's not so terrible here, plus we could help with the comprehension via a consistent syntax highlighting across plugins. Still imho this is "big" enough to try to stick to a universal syntax instead. Since this seems to work for Grafana, why not just do the same?

**Reply by saminzadeh** [reply]
For the time being, I think we should _suggest_  the `$` prefix syntax when consuming variables within a string if you don't have something already, but not require it.

However, I think the exact syntax to **consume** a variable is out of scope for this task. Instead, we should just focus on the name of the variable. In that regard, we can reserve the `__` prefix for built-in variable names (either at dashboard scope or plugin scope) meaning user-defined variable names cannot contain the `__` prefix.

eg. The variable `__start_time_ms` can be consumed in one plugin via `$__start_time_ms` and `@__start_time_ms` in another.

**Reply by AntoineThebaud** [reply]
I agree for the `__` prefix being reserved for built-in variables.

And indeed let's create a separate discussion for this variables syntax topic. EDIT it's available [here](https://github.com/perses/perses/discussions/1234).
